Data/DataModel/Puzzle.cs

```diff
@@ -43,34 +46,70 @@ public class Puzzle
         /// </summary>
         public int MinPrerequisiteCount { get; set; } = 0;
 
-        [DataType(DataType.Url)]
-        public string PuzzleUrlString { get; set; }
+        /// <summary>
+        /// All of the content files associated with this puzzle
+        /// </summary>
+        public virtual ICollection<ContentFile> Contents { get; set; }
 
+        /// <summary>
+        /// File for the main puzzle (typically a PDF containing the puzzle)
+        /// </summary>
         [NotMapped]
-        public Uri PuzzleUrl
+        public ContentFile PuzzlePDF
         {
-            get { Uri.TryCreate(PuzzleUrlString, UriKind.RelativeOrAbsolute, out Uri result); return result; }
-            set { PuzzleUrlString = value?.ToString(); }
+            get
+            {
+                var puzzlePDFs = from contentFile in Contents ?? Enumerable.Empty<ContentFile>()
+                                 where contentFile.FileType == ContentFileType.Puzzle
+                                 select contentFile;
+                Debug.Assert(puzzlePDFs.Count() <= 1);
+                return puzzlePDFs.FirstOrDefault();
+            }
         }
 
-        [DataType(DataType.Url)]
-        public string AnswerUrlString { get; set; }
-
+        /// <summary>
+        /// File for the main answer (typically a PDF containing the answer)
+        /// </summary>
         [NotMapped]
-        public Uri AnswerUrl
+        public ContentFile AnswerPDF
         {
-            get { Uri.TryCreate(AnswerUrlString, UriKind.RelativeOrAbsolute, out Uri result); return result; }
-            set { AnswerUrlString = value?.ToString(); }
+            get
+            {
+                var answerPDFs = from contentFile in Contents ?? Enumerable.Empty<ContentFile>()
+                                 where contentFile.FileType == ContentFileType.Answer
+                                 select contentFile;
+                Debug.Assert(answerPDFs.Count() <= 1);
+                return answerPDFs.FirstOrDefault();
+            }
         }
 
-        [DataType(DataType.Url)]
-        public string MaterialsUrlString { get; set; }
 
+        /// <summary>
+        /// Files associated with the puzzle, such as media
+        /// </summary>
         [NotMapped]
-        public Uri MaterialsUrl
+        public IEnumerable<ContentFile> Materials
```

Comment:
Just curious - can you build a rudimentary website out of Materials? Does that work? I know that you can't possibly have backend code, really I'm just asking about whether you can make a simple HTML page that links in lots of images/audio.

Reply:
We should allow zip files that get unpacked into storage blobs (as well allowing ones that don't). I'll think about a way to do that in a future change. (#65)